File: lib/resource/infoSection.ts
import { ByteReader } from './byteReader'
import {
  DatId,
  InfoResource,
  type InfoDefinition,
  type MountDefinition,
  movementTypeFromIndex,
  rangeTypeFromIndex,
  RangeType,
} from './datResource'
import { ParserResult, type ResourceParser, type SectionHeader } from './datParser'

const TWO_PI = Math.PI * 2

export class InfoSection implements ResourceParser {
  private readonly sectionHeader: SectionHeader
  private infoDefinition: InfoDefinition = {
    movementType: movementTypeFromIndex(0xff),
    movementChar: '0',
    shakeFactor: 0,
    weaponAnimationType: 0,
    weaponAnimationSubType: 0,
    standardJointIndex: null,
    scale: null,
    staticNpcScale: null,
    rangeType: RangeType.Unset,
  }

  private mountDefinition: MountDefinition = {
    rotation: 0,
    poseType: 0,
  }

  constructor(sectionHeader: SectionHeader) {
    this.sectionHeader = sectionHeader
  }

  getResource(byteReader: ByteReader): ParserResult {
    this.read(byteReader)
    return ParserResult.from(new InfoResource(this.sectionHeader.sectionId, this.infoDefinition, this.mountDefinition))
  }

  private read(byteReader: ByteReader): void {
    if (this.sectionHeader.sectionId.id === DatId.mount.id) {
      this.readMountDefinition(byteReader)
      return
    }

    this.readInfoDefinition(byteReader)
  }

  private readInfoDefinition(byteReader: ByteReader): void {
    byteReader.offsetFromDataStart(this.sectionHeader, 0)

    const movementType = movementTypeFromIndex(byteReader.next8())
    const movementChar = this.toMovementChar(byteReader.next8())
    const shakeFactor = byteReader.next8()

    const weaponAnimationType = byteReader.next8()
    const weaponAnimationSubType = byteReader.next8()
    byteReader.next8()
    const standardJointIndex = this.nullIf0xff(byteReader.next8())

    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    const scale = this.nullIf0xff(byteReader.next8())
    const staticNpcScale = this.nullIf0xff(byteReader.next8())
    byteReader.next8()
    byteReader.next8()

    const rawRangeType = byteReader.next8()
    byteReader.next8()

    this.infoDefinition = {
      movementType,
      movementChar,
      shakeFactor,
      weaponAnimationType,
      weaponAnimationSubType,
      standardJointIndex,
      scale,
      staticNpcScale,
      rangeType: rangeTypeFromIndex(rawRangeType) ?? RangeType.Unset,
    }
  }

  private readMountDefinition(byteReader: ByteReader): void {
    byteReader.offsetFromDataStart(this.sectionHeader, 0)
    byteReader.next8()
    byteReader.next8()
    const rotation = byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    const poseType = byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()
    byteReader.next8()

    this.mountDefinition = {
      rotation: TWO_PI * (rotation / 255),
      poseType,
    }
  }

  private toMovementChar(value: number): string {
    return value === 0xff ? '0' : DatId.base36ToChar(value)
  }

  private nullIf0xff(value: number): number | null {
    return value === 0xff ? null : value
  }
}

File: xim/resource/InfoSection.kt
package xim.resource

import xim.util.OnceLogger
import xim.util.PI_f

enum class RangeType(val index: Int, val subtype: Int) {
    None(0x00, 1),
    Wind(0x01, 1),
    String(0x02, 1),
    Marksmanship(0x03, 0),
    ThrowingWeapon(0x04, 0), // Boomerang-type?
    ThrowingAmmo(0x05, 0), // Boomerang-type?
    Archery(0x06, 0),
    // no 0x07
    // no 0x08
    // no 0x09
    HandbellIndi(0x0A, 2),
    HandbellGeo(0x0B, 2),
    Unset(0xFF, 0),
    ;

    companion object {
        fun fromIndex(index: Int) : RangeType? {
            return RangeType.values().firstOrNull { it.index == index }
        }
    }

}

enum class MovementType(val index: Int, val slopeOriented: Boolean = false) {
    Walking(0x0),
    Sliding(0x1, slopeOriented = true), //??? Pugil, Crawler, etc
    Large(0x2, slopeOriented = true), //??? Morbol, Treant, Behe, SeaMonk
    Flying(0x3), // bats, birds, bees, bombs, etc
    Unset(0xFF), // NPCs that don't move, footwear, etc
    ;

    companion object {
        fun fromIndex(index: Int) : MovementType {
            return MovementType.values().firstOrNull { it.index == index } ?: throw IllegalStateException("Unknown movement type: ${index.toString(0x10)}")
        }
    }
}

class InfoDefinition(
    val movementType: MovementType = MovementType.Unset,
    val movementChar: Char = '0',
    val shakeFactor: Int = 0,
    val weaponAnimationType: Int = 0,
    val weaponAnimationSubType: Int = 0,
    val standardJointIndex: Int? = null,
    val scale: Int? = null,
    val staticNpcScale: Int? = null,
    val rangeType: RangeType = RangeType.Unset,
)

class MountDefinition(
    val rotation: Float = 0f,
    val poseType: Int = 0,
)

class InfoSection(private val sectionHeader: SectionHeader) : ResourceParser {

    private var infoDefinition: InfoDefinition = InfoDefinition()
    private var mountDefinition: MountDefinition = MountDefinition()

    override fun getResource(byteReader: ByteReader): ParserResult {
        read(byteReader)
        val resource = InfoResource(sectionHeader.sectionId, infoDefinition, mountDefinition)
        return ParserResult.from(resource)
    }

    private fun read(byteReader: ByteReader) {
        when (sectionHeader.sectionId) {
            DatId.info -> readInfoDefinition(byteReader)
            DatId.mount -> readMountDefinition(byteReader)
            else -> {
                OnceLogger.error("[$byteReader] Unknown info type: $sectionHeader. Defaulting to standard info type.")
                readInfoDefinition(byteReader)
            }
        }
    }

    private fun readInfoDefinition(byteReader: ByteReader) {
        byteReader.offsetFromDataStart(sectionHeader, 0x0)

        val movementType = MovementType.fromIndex(byteReader.next8())
        val movementChar = toMovementChar(byteReader.next8())
        val shakeFactor = byteReader.next8()    // Not sure

        val weaponType = byteReader.next8()
        val weaponSubType = byteReader.next8()
        val weaponUnk1 = byteReader.next8()
        val standardJointIndex = nullIf0xFF(byteReader.next8())

        val unk0x07 = byteReader.next8()
        val unk0x08 = byteReader.next8()
        val unk0x09 = byteReader.next8()
        val scale = nullIf0xFF(byteReader.next8())
        val staticNpcScale = nullIf0xFF(byteReader.next8())
        val unk0x0C = byteReader.next8()
        val unk0x0D = byteReader.next8()
        val rawRangeType = byteReader.next8()
        val unk0x0F = byteReader.next8()

        val rangeType = RangeType.fromIndex(rawRangeType)
        if (rangeType == null) { OnceLogger.warn("[$byteReader] Unknown range type: ${rawRangeType.toString(0x10)}") }

        infoDefinition = InfoDefinition(
            movementType = movementType,
            movementChar = movementChar,
            shakeFactor = shakeFactor,
            weaponAnimationType = weaponType,
            weaponAnimationSubType = weaponSubType,
            standardJointIndex = standardJointIndex,
            scale = scale,
            staticNpcScale = staticNpcScale,
            rangeType = rangeType ?: RangeType.Unset,
        )
    }

    private fun readMountDefinition(byteReader: ByteReader) {
        val unk0x00 = byteReader.next8()
        val unk0x01 = byteReader.next8()
        val rotation = byteReader.next8()
        val unk0x03 = byteReader.next8()
        val unk0x04 = byteReader.next8()
        val unk0x05 = byteReader.next8()
        val unk0x06 = byteReader.next8()
        val unk0x07 = byteReader.next8()
        val unk0x08 = byteReader.next8()
        val unk0x09 = byteReader.next8()
        val poseType = byteReader.next8()
        val unk0x0B = byteReader.next8()
        val unk0x0C = byteReader.next8()
        val unk0x0D = byteReader.next8()
        val unk0x0E = byteReader.next8()
        val unk0x0F = byteReader.next8()

        mountDefinition = MountDefinition(
            rotation = 2 * PI_f * (rotation / 255f),
            poseType = poseType,
        )
    }

    private fun toMovementChar(value: Int): Char {
        return if (value == 0xFF) { '0' } else { DatId.base36ToChar(value) }
    }

    private fun nullIf0xFF(value: Int): Int? {
        return if (value == 0xFF) { null } else { value }
    }

}
